ingestion-core/src/main/java/com/mozilla/telemetry/ingestion/core/transform/PubsubMessageToObjectNode.java

```diff
@@ -595,6 +650,76 @@ private String getAndCacheBqName(String name) {
       }
     }
 
+    /**
+     * Encode the given JSON object as a compact string.
+     *
+     * <p>We maintain a BigQuery persistent UDF that can parse a variety of formats, so any format
+     * used here must first be supported in:
+     * https://github.com/mozilla/bigquery-etl/blob/master/mozfun/hist/extract/udf.sql
+     *
+     * <p>Schema validation should generally ensure that all histograms are well-formed, but we
+     * perform some light sanity checking on the values and fall back to encoding the histogram
+     * as a JSON string if anything seems off.
+     */
+    private TextNode compactHistogramEncoding(JsonNode o, String fieldName) {
+      final int histogramType = o.path("histogram_type").asInt(-1);
+      if (histogramType < 0 || histogramType > 5) {
+        incrementInvalidHistogramType();
+        return jsonHistogramEncoding(o);
+      }
+      final long sum = o.path("sum").asLong(-1);
+      if (sum < 0) {
+        incrementInvalidHistogramSum();
+        return jsonHistogramEncoding(o);
+      }
+      final JsonNode values = o.path("values");
+      if (histogramType == 2 && fieldName.startsWith("use_counter2")) {
+        // Histograms named as "use_counter" are reported as type 2 (boolean), but only ever have
+        // a non-zero value in the "1" (true) bucket (and the "sum" field should match this count).
+        // They can be encoded as a textual representation of that single number without any loss
+        // of information, and since use counters make up the majority of histogram data volume,
+        // this optimization case is the most important one.
+        // See https://firefox-source-docs.mozilla.org/toolkit/components/telemetry/collection/use-counters.html
+        if (sum == values.path("1").asLong(-1)) {
+          return TextNode.valueOf(Long.toString(sum));
+        } else {
+          incrementInvalidHistogramUseCounter();
```

Comment:
I was able to track down that all the "invalid" use counters we found were ones that had sum = 0 and an empty values map. These invalids all go away if we catch this case and write out 0 instead. But I think a use counter with value 0 is semantically equivalent to null, so I am considering returning a null node in the case of `sum = 0`.

Reply:
I've pushed an additional commit so that we now encode these as "0".